jlm/util/BiMap.hpp

```diff
@@ -0,0 +1,254 @@
+ * @tparam V the type of values
+ */
+template<typename K, typename V>
+class BiMap
+{
+  /**
+   * Adds a key/value pair to only the reverse map.
+   * @throws jlm::util::error if the value already exists in the BiMap.
+   */
+  void
+  AddPairToReverse(K key, V value)
+  {
+    auto it = ReverseMap_.find(value);
+    if (it != ReverseMap_.end())
+      throw jlm::util::error("Value already exsists in BiMap");
+    ReverseMap_.insert({ value, key });
+  }
+
+public:
+  ~BiMap() noexcept = default;
+
+  BiMap() = default;
+
+  BiMap(const BiMap & other) = default;
+
+  BiMap(BiMap && other) noexcept = default;
+
+  BiMap &
+  operator=(const BiMap & other) = default;
+
+  BiMap &
+  operator=(BiMap && other) = default;
+
+  /**
+   * Constructs a BiMap from an existing unordered_map.
+   * @throws jlm::util::error if the values are not unique
+   */
+  explicit BiMap(const std::unordered_map<K, V> & forwardMap)
+      : ForwardMap_(forwardMap),
+        ReverseMap_()
+  {
+    for (auto & [key, value] : ForwardMap_)
+      AddPairToReverse(key, value);
+  }
+
+  /**
+   * Removes all key/value mappings from the BiMap.
+   */
+  void
+  Clear() noexcept
+  {
+    ForwardMap_.clear();
+    ReverseMap_.clear();
+  }
+
+  /**
+   * Gets the number of key/value mappings in the BiMap.
+   */
+  [[nodiscard]] std::size_t
+  Size() const noexcept
+  {
+    return ForwardMap_.size();
+  }
+
+  /**
+   * Inserts a new key/value pair into the BiMap,
+   * @param key the key to be inserted, with the associated value
+   * @param value the value to be inserted, associated with the key
+   * @throws jlm::util::error if the key or value already exists in the BiMap
+   */
+  void
+  Insert(const K & key, const V & value)
+  {
+    auto it = ForwardMap_.find(key);
+    if (it != ForwardMap_.end())
+      throw jlm::util::error("Key already exsists in BiMap");
+    AddPairToReverse(key, value);
+    ForwardMap_.insert({ key, value });
+  }
+
+  /**
+   * Checks for the existance of the given \p key in the BiMap
+   * @param key the key to look for
+   * @returns true if the key exists in the BiMap, otherwise false.
+   */
+  [[nodiscard]] bool
+  HasKey(const K & key) const noexcept
+  {
+    return ForwardMap_.find(key) != ForwardMap_.end();
+  }
+
+  /**
+   * Checks for the existance of the given \p value in the BiMap
+   * @param value the value to look for
+   * @returns true if the value exists in the BiMap, otherwise false.
+   */
+  [[nodiscard]] bool
+  HasValue(const V & value) const noexcept
+  {
+    return ReverseMap_.find(value) != ReverseMap_.end();
+  }
+
+  /**
+   * @param key the key to lookup in the BiMap
+   * @returns the value associated with the given \p key
+   * @throws jlm::util::error if the \p key does not exist in the BiMap
+   */
+  [[nodiscard]] const V &
+  LookupKey(const K & key) const
+  {
+    auto it = ForwardMap_.find(key);
+    if (it == ForwardMap_.end())
+      throw jlm::util::error("Key not found in BiMap");
```

Comment:
The same error message is used multiple times in the code. Could we refactor it out to a private static string in the class?